src/libraries/System.Text.Json/gen/src/Reflection/MetadataLoadContextInternal.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis;
+
+namespace System.Text.Json.SourceGeneration.Reflection
+{
+    internal class MetadataLoadContextInternal
+    {
+        private readonly Dictionary<string, IAssemblySymbol> _assemblies = new Dictionary<string, IAssemblySymbol>(StringComparer.OrdinalIgnoreCase);
+
+        private readonly Compilation _compilation;
+
+        private IAssemblySymbol? _collectionsAssemblySymbol;
+
+        public MetadataLoadContextInternal(Compilation compilation)
+        {
+            _compilation = compilation;
+            Dictionary<AssemblyName, IAssemblySymbol> assemblies = compilation.References
```

Comment:
Did you see any major performance regression since this happens every build?

Reply:
I really haven't done much build-time throughput testing, but [really] early code-gen perf [numbers](https://gist.github.com/layomia/25ce825e2f9abc22e49b344edae65601) suggested the build time isn't increasing by a crazy amount. Do you know of a reliable way to measure the perf here that doesn't include noise from other parts of the compilation?

I've taken a [note](https://github.com/dotnet/runtime/issues/45448#issuecomment-822144624) to optimize the generator's implementation (e.g. looking at caching where possible - like this code snippet maybe?), and also to adopt the incremental source generator pattern being designed.